server/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java

```diff
@@ -65,6 +67,17 @@ public String getName() {
         return name;
     }
 
+    /**
+     * Return the field names this aggregation creates.
+     *
+     * This method is a optional helper for clients that need to know the output field names.
+     *
+     * @return The set of output field names this aggregation produces or Optional.empty() if unknown or not implemented.
```

Comment:
@nik9000 did you want to distinguish between "not implemented" and "we can't know the output fields", using null for one of them?

Reply:
I want to make sure we can tell the difference between "I didn't think about pipelines at all when implementing this agg" and "there are no fields exposed by this agg for pipelines". I think `Optional.empty` can be "I didn't think about it" and "Optional.of(emptySet())` can be "there are no fields for pipelines here".

Reply:
```suggestion
     * @return The set of output field names this aggregation produces or Optional.empty() if not implemented or Optional.of(emptySet()) if the fields are not known.
```
Maybe something like this then?